### PR TITLE
tests from ormbattle #573

### PR DESCRIPTION
**Pull request by jogibear9988**

### CODE REVIEW

**Comment by jogibear9988** [comment]
why? i think it's good if all this would be tested, so we'll see if smth will break linq functionality

**Comment by jogibear9988** [comment]
I think it's a good idea, to add all the tests. Look at test "WhereNullableParameterTest", it's broken cause of a change we did in sqlite with datetime, so we would not find out if we only add tests wich were broken.

But I also think, we should add the tests for all DB Providers, not only sqlite, but that can be a step 2

File: Source/DataProvider/SQLite/SQLiteSqlBuilder.cs [hunk 1]
```diff
@@ -131,17 +131,22 @@ protected override void BuildPredicate(ISqlPredicate predicate)
 				var leftType  = exprExpr.Expr1.SystemType;
 				var rightType = exprExpr.Expr2.SystemType;
 
-				if (IsDateTime(leftType) || IsDateTime(rightType))
+				if ((IsDateTime(leftType) || IsDateTime(rightType)) && !((exprExpr.Expr1 is SqlValue && ((SqlValue)exprExpr.Expr1).Value == null) || (exprExpr.Expr2 is SqlValue && ((SqlValue)exprExpr.Expr2).Value == null)))
```
Comment (ili) [inline Source/DataProvider/SQLite/SQLiteSqlBuilder.cs]: @jogibear9988 it is better to check if type is `IValueContainer`, this way you will close two cases: `SqlValue` & `SqlParameter`

File: Source/DataProvider/SQLite/SQLiteSqlBuilder.cs [hunk 2]
```diff
@@ -131,17 +131,22 @@ protected override void BuildPredicate(ISqlPredicate predicate)
 				var leftType  = exprExpr.Expr1.SystemType;
 				var rightType = exprExpr.Expr2.SystemType;
 
-				if (IsDateTime(leftType) || IsDateTime(rightType))
+				if ((IsDateTime(leftType) || IsDateTime(rightType)) && !((exprExpr.Expr1 is SqlValue && ((SqlValue)exprExpr.Expr1).Value == null) || (exprExpr.Expr2 is SqlValue && ((SqlValue)exprExpr.Expr2).Value == null)))
 				{
+					
+					if (leftType != null)
```
Comment (ili) [inline Source/DataProvider/SQLite/SQLiteSqlBuilder.cs]: @jogibear9988 `leftType` or `rightType` would be `null` for `null` `Value`?

File: Source/Linq/Builder/ExpressionBuilder.cs [hunk 3]
```diff
@@ -421,8 +421,13 @@ TransformInfo OptimizeExpressionImpl(Expression expr)
 							var isList = typeof(ICollection).IsAssignableFromEx(me.Member.DeclaringType);
 
 							if (!isList)
+								isList =
```
Comment (ili) [inline Source/Linq/Builder/ExpressionBuilder.cs]: @jogibear9988 what does it for? We do check the same below and in more general way. I think this check is not needed.

**Comment by ili** [comment]
Speaking about OrmBattle tests, i'm agree with @sdanyliv - it is better to rewrite failing tests using our object model, and run them against **all databases** - for ex "WhereNullableParameterTest" fails because of bug in SQLite DateTime compassion, and we are not sure that it is only database where it fails.

So, my idea is to treat tests failing now as point to fix now, and check for all databases. In future may be we'll rewrite all tests using our model to run for all databases...

Or, may be we would migrate to https://github.com/lerocha/chinook-database instead of Northwind

File: Source/Linq/Builder/ExpressionBuilder.cs [hunk 4]
```diff
@@ -421,8 +421,13 @@ TransformInfo OptimizeExpressionImpl(Expression expr)
 							var isList = typeof(ICollection).IsAssignableFromEx(me.Member.DeclaringType);
 
 							if (!isList)
+								isList =
+									me.Member.DeclaringType.IsGenericType &&
```
Comment (ili) [inline Source/Linq/Builder/ExpressionBuilder.cs]: @jogibear9988 one more point - there are reflection extensions (in `ReflectionExtensions.cs`) they are written to be compatible for different frameworks (conditional compilation), so in future avoid using direct reflection functions, firstly check if there are one in extensions.

For ex if you want to chech if type is generic use `ReflectionExtensions.IsGenericTypeEx(this Type type)`.
Most methods are with the same name as reflection methods or properties, only having `Ex` postfix :)

File: Source/DataProvider/SQLite/SQLiteSqlBuilder.cs (hunk 2 above)
Comment (jogibear9988) [inline Source/DataProvider/SQLite/SQLiteSqlBuilder.cs]: it is in a test from ormbattle

File: Source/DataProvider/SQLite/SQLiteSqlBuilder.cs (hunk 1 above)
Comment (jogibear9988) [inline Source/DataProvider/SQLite/SQLiteSqlBuilder.cs]: don't know what you mean

**Comment by jogibear9988** [comment]
i'll rewrite the tests to use our model, when the errors are fixed

File: Source/Linq/Builder/ExpressionBuilder.cs (hunk 3 above)
Comment (jogibear9988) [inline Source/Linq/Builder/ExpressionBuilder.cs]: with my change test works, without not...

File: Source/Linq/Builder/ExpressionBuilder.cs (hunk 4 above)
Comment (jogibear9988) [inline Source/Linq/Builder/ExpressionBuilder.cs]: NetStandart will have the same reflection methods in 2.0 as full framework. Du we then still need the extensions?

File: Source/Linq/Builder/ExpressionBuilder.cs (hunk 3 above)
Comment (jogibear9988) [inline Source/Linq/Builder/ExpressionBuilder.cs]: if you think the check 2 lines lower works, it does not if your type is directly icollection<T>, which is the case here!

**Comment by jogibear9988** [comment]
Some of the OrmBattle Test will not work on Sqlite on NetCore, see: https://github.com/aspnet/Microsoft.Data.Sqlite/issues/331

**Comment by jogibear9988** [comment]
I've switched the OrmBattle Tests to our model

**Comment by jogibear9988** [comment]
have you looked at my reworked fixes? how should I go on with this?

**Comment by ili** [comment]
I'll take a look today :)

**Comment by ili** [comment]
>Some of the OrmBattle Test will not work on Sqlite on NetCore, see: aspnet/Microsoft.Data.Sqlite#331

yes. they do say that it is because MS uses official SQLite build, and System.Data.SQLite - non official

That's why i'v introduced SQLiteMs configuration for tests, exclude failing tests.

File: Source/Linq/Builder/ExpressionBuilder.cs (hunk 3 above)
Comment (ili) [inline Source/Linq/Builder/ExpressionBuilder.cs]: `me.Member.DeclaringType.GetInterfacesEx()` do not return `ICollection<>`?? that's strange...

**Comment by ili** [comment]
>have you looked at my reworked fixes? how should I go on with this?

Yes! :)

**Comment by jogibear9988** [comment]
yes?

i do mean, i'm not ableto fix the rest of the tests.. (the sqlite ones i will change to sqlitems). should i disable them ? so this could be merged? 

we could kerp the issue open, so we know we've work to do...

**Comment by ili** [comment]
Some misunderstood. I mean that you should continue working on this.

Yes, we can disable some tests for .net core using conditional compilation.

Other failing test may be disabled for now

**Comment by jogibear9988** [comment]
I've now added

     if (context != TestProvName.SQLiteMs)
			{

to the failing tests... but they still fail...

**Comment by jogibear9988** [comment]
Okay, we only have "NorthwindSqlite" and "Northwind". But how can I check if "NorthwindSqlite" uses Microsoft.sqlite?

**Comment by ili** [comment]
> Okay, we only have "NorthwindSqlite" and "Northwind". But how can I check if "NorthwindSqlite" uses Microsoft.sqlite?

New configuration `NorthwindSqliteMs` is needed. So for Core we'll use `NorthwindSqliteMs` instead of `NorthwindSqlite` in `UserDataProviders.Core.txt`

Then we'll need to make some rewrite of `NorthwindDataContextAttribute`, like this:

```cs
[AttributeUsage(AttributeTargets.Method)]
public class NorthwindDataContextAttribute : IncludeDataContextSourceAttribute
{
	public NorthwindDataContextAttribute(bool excludeSqlite, bool excludeSqliteMs = false) : base(
		excludeSqlite
		? new[] { "Northwind" }
		:  ( excludeSqliteMs  ? new[] { "Northwind", "NorthwindSqlite" } : new[] { "Northwind", "NorthwindSqlite",  "NorthwindSqliteMs"}))
	{
	}

	public NorthwindDataContextAttribute() : this(false)
	{
	}
}
``` 

and will mark excluding tests as `[NorthwindDataContext(false, true)]`

**Comment by jogibear9988** [comment]
i will do so

**Comment by jogibear9988** [comment]
the tests now work ;-)

**Comment by jogibear9988** [comment]
can be merged....

**Comment by jogibear9988** [comment]
but don't close #573, there are still failing tests

**Pull request merged**